File: 1423-maximum-points-you-can-obtain-from-cards/1423-maximum-points-you-can-obtain-from-cards.cpp
class Solution {
public:
    int maxScore(vector<int>& cardPoints, int k) {
        // Approach_02: Using Prefix Sum: Time:(2k) and Space:O(n)
        int n = cardPoints.size();
        vector<int>left(k+1), right(k+1);
        left[0] = 0, right[0] = 0;
        
        for(int i = 0;i<k;i++)
        {
            left[i+1] = cardPoints[i]+left[i];
            right[i+1] = cardPoints[n-i-1]+right[i];
        }
        int mx = 0;
        for(int i = 0;i<=k;i++)
        {
            int cur = left[i]+right[k-i];
            mx = max(mx,cur);
        }
        return mx;
    }
};

File: 1423-maximum-points-you-can-obtain-from-cards/1423-maximum-points-you-can-obtain-from-cards.java
class Solution {
    public int maxScore(int[] cardPoints, int k) {
        //Approach_03: Using Sliding window Technique: Time:O(k) and Space:O(1)
        int sum = 0, n = cardPoints.length;
        
        // step1: Sum of first k elements: i.e sum of left window;
        for(int i = 0;i<k;i++){
            sum += cardPoints[i];
        }
        // Step2: substract the value from k-1 till 0 and add the value cardPoints[n+i-k]
        int maxPoints = sum;
        for(int i = k-1;i>=0;i--)
        {
            sum = sum - cardPoints[i]+cardPoints[n+i-k];
            maxPoints = Math.max(maxPoints, sum);
        }
        return maxPoints;
    }
}